### PR TITLE
Replace axios with (node-)fetch

### PR DESCRIPTION
**Pull request by janpaepke**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -51,7 +51,7 @@
     "@mollie/eslint-config-typescript": "^1.6.5",
     "@types/jest": "^29.5.11",
     "@types/node": "^18.14.6",
-    "axios-mock-adapter": "1.19.0",
+    "@types/node-fetch": "^2.6.11",
```
Comment (Pimm) [inline package.json:54]: The `@types` and the regular dependency are out of sync.

Comment (janpaepke) [reply package.json:54]: It's the same major and at the time of writing there is no newer types package for node-fetch.
This should be fine.

File: src/Options.ts [hunk 2]
```diff
@@ -29,7 +27,7 @@ type Options = Xor<
    * The URL of the root of the Mollie API. Default: `'https://api.mollie.com:443/v2/'`.
    */
   apiEndpoint?: string;
-} & Pick<AxiosRequestConfig, 'adapter' | 'proxy' | 'socketPath' | 'timeout'>;
+};
```
Comment (Pimm) [inline src/Options.ts:30]: We should document that this is a breaking change.

**Comment by Pimm** [comment]
Fixes https://github.com/mollie/mollie-api-node/issues/346.

**Review by fjbender — APPROVED** [review]

File: src/communication/NetworkClient.ts [hunk 3]
```diff
@@ -69,21 +68,37 @@ const throwApiError = run(
...
 interface Data {}
 interface Context {}
 
 /**
- * This class is essentially a wrapper around axios. It simplifies communication with the Mollie API over the network.
+ * This class is essentially a wrapper around fetch. It simplifies communication with the Mollie API over the network.
  */
 export default class NetworkClient {
-  protected readonly axiosInstance: AxiosInstance;
+  protected readonly request: (pathname: string, options?: RequestInit) => Promise<ResponseWithIdempotencyKey>;
```
Comment (Pimm) [inline src/communication/NetworkClient.ts:101]: Could we add one or two sentences to explain what this function does, specifically how it differs from (vanilla) `fetch`?

File: src/communication/NetworkClient.ts [hunk 4]
```diff
@@ -92,71 +107,67 @@ export default class NetworkClient {
...
-      return true;
-    }
-    return response.data;
+    const { idempotencyKey, ...body } = data;
+    const config: RequestInit = {
+      method: 'POST',
+      headers: idempotencyKey ? { [idempotencyHeaderName]: idempotencyKey } : undefined,
+      body: JSON.stringify(body),
+    };
+    return this.request(buildUrl(pathname, query), config).catch(throwApiError).then(processFetchResponse);
```
Comment (Pimm) [inline src/communication/NetworkClient.ts:148]: The `.catch(throwApiError).then(processFetchResponse)` seems repetitive. Can we change the abstraction somehow?

File: src/communication/NetworkClient.ts [hunk 5]
```diff
@@ -69,21 +68,37 @@ const throwApiError = run(
...
+async function processFetchResponse(response: ResponseWithIdempotencyKey) {
+  // Request was successful, but no content was returned.
+  if (response.status == 204) return true;
+  // Request was successful and content was returned.
+  const body = await response.json();
+  if (Math.floor(response.status / 100) == 2) {
+    return body;
+  }
+  // Request was not successful, but the response body contains an error message.
+  if (body) {
```
Comment (Pimm) [inline src/communication/NetworkClient.ts:87]: Can we replace this truth-y check with something else? Are we checking for `undefined` here? Or null-ish?

File: src/communication/makeRetrying.ts [hunk 6]
```diff
@@ -42,83 +31,69 @@ function generateIdempotencyHeader() {
...
   }
   // Convert the header value from seconds to milliseconds.
   return retryAfter * 1e3;
 }
 
+export type ResponseWithIdempotencyKey = Response & { idempotencyKey: string | undefined };
+
 /**
- * Enhances the passed Axios instance, making it attempt requests multiple times in some scenarios.
+ * Wrapper around fetch, making it attempt requests multiple times in some scenarios.
```
Comment (Pimm) [inline src/communication/makeRetrying.ts:56]: Could we change this documentation to start with a verb?

File: tests/unit/resources/__snapshots__/lists.test.ts.snap [hunk 7]
```diff
@@ -64,68 +64,3 @@ exports[`lists .list() should retrieve the next page 1`] = `
   },
 ]
 `;
-
-exports[`lists .list() should retrieve the next page 2`] = `
```
Comment (Pimm) [inline tests/unit/resources/__snapshots__/lists.test.ts.snap:68]: Could we double-check that this is intentional?

Comment (janpaepke) [reply tests/unit/resources/__snapshots__/lists.test.ts.snap:68]: Yes, this is indeed intentional, as there were previously two tests doing the same thing and they have been refactored to be one single test, rendering one of the snapshots obsolete.

**Comment by Pimm** [comment]
### A note on Node.js version support.

One could make the argument that periodically increasing the Node.js version this client requires is a good thing, because it motivates users to upgrade their Node.js version to one which is still in the maintenance window. I personally suspect and fear the true effect would instead be that users stop updating this client altogether. I therefore continue to believe the best strategy is supporting all Node.js versions which are still being maintained, as well as any older version we can reasonably support – including through polyfills.

The current oldest Node.js version which is still being maintained is 18, which will be maintained until 30 April 2025. However, through [node-fetch](https://github.com/node-fetch/node-fetch), we can easily support 14 and up.

In fact, node-fetch 2.× supports Node.js all the way down to 6. However, Node.js versions under 14 are challenging to test, as Jest and Nock have version requirements of their own.

We decided to raise the Node.js version requirement to 14, while _secretly_ supporting 8. We've tested the client including this PR on Node.js 8, and it does work, but the test suite itself needs to be modified to make the tests themselves compatible.

**Pull request merged**